source/common/network/connection_impl.cc

```diff
@@ -224,6 +224,11 @@ void ConnectionImpl::enableHalfClose(bool enabled) {
 
 void ConnectionImpl::readDisable(bool disable) {
   ASSERT(state() == State::Open);
+  if (state() != State::Open || file_event_ == nullptr) {
```

Comment:
The following is for philosophical discussion, not exactly related to this PR:

What is the role of an ASSERT? I my mind, it's something that shouldn't happen, and if it does happen, we're in such an unexpected state that the best thing to do is to abort(), collect a core file, and send it to the developers to figure out how to make it not happen.

Writing an assert, and then handling the case of it being false, seems like the wrong thing. In my opinion, the ASSERT should be removed here.

I'm also in favor of ASSERT being compiled into release builds, but that's a discussion for another day.

Reply:
I think you're looking for RELEASE_ASSERT

I believe ASSERT is more for bounds checking, invariant checking etc.  I think in the case where the watermark code is called on a closed connection, we're (IMO) better off running than crashing.  We could remove the ASSERT here but I think it is an unintended bug that the tcp proxy code is applying watermark limits to a closed connection.  I just think error handling is better than crashing in this case.  If we think it should be a legal non-bug to disable reads on a closed connection I could take it out, but that does seem conceptually weird to me.

Reply:
Ah yeah, this is documented in STYLE.md

Reply:
I guess what I find weird is that we're declaring an invariant (with the assert), and then handling the case where the invariant doesn't hold, thus implying it's not an invariant.

I agree that error handling is better than crashing here. But if we add error handling for a condition, I don't feel like there should be an ASSERT that the condition doesn't happen.

Reply:
From STYLE.md

"At a very high level, our philosophy is that errors that are likely to happen should be gracefully handled. Examples of likely errors include any type of network error, disk IO error, bad data returned by an API call, bad data read from runtime files, etc."

"ASSERT(condition) should be used to document (and check in debug-only builds) program invariants."

I believe that calling readDisable on a closed connection is an error.  To me, that there is logically a program invariant on calling readDisable on a closed connection.  ASSERTS are encouraged for documenting invariants. This PR results in us handling it gracefully, as we are encouraged to do with errors.   I see no conflict between noting something should never happen and gracefully handling it if it does happen, and it appears to me to be consistent with our style guide.  You can totally argue we should make changes to the style guide but I'd prefer they not block this PR :-)

Reply:
Yes, I agree, no need to block this PR. Maybe I'll create #philosophy in slack.

Reply:
Haha, I'll totally join that channel.

FWIW I think Matt and Harvey tend to lean towards the hard-fail school as well.  I push pretty hard on the recover-if-at-all-humanly-possible because it's reaaaaally crummy when a query of death takes out all of google.com because we did the equivalent of a RELEASE_ASSERT instead of ASSERT on some previously unexercised code path :-/

For the long term, I'm hoping to port our internal bug handler to Envoy, and replace many of our ASSERTS with it.  It basically crashes in debug mode as ASSERT does (ideally catches bugs in integration tests), logs details (including a stack trace and line numbers) to error log in opt mode*, and increments a mapped stat with line number which we then monitor.  When we hit things like this in production we error handle, but we can also use the global monitoring to track down instances, retrieve logs, and keep iterating on the error logs until we can reproduce and fixed it.  Best of both worlds, I think. 

*with power of 2 backoff, so as to give you data as to how much it's happening without overwhelming your process with immediately flushed error logs.

Reply:
That sounds awesome. I'd just been looking (during this discussion) at whether we could make ASSERT log (but not crash/PANIC) in release builds, so at least we have some indication that these cases are happening in production.

Reply:
FWIW I agree with @ggreenway here. This seems not right, in the sense that we should make the assert not happen ever, or remove the assert. Are we tracking this in a follow up anywhere?